### ISSUE TITLE
Edges in val_edges and test_edges are trained with label '0'?

### ISSUE THREAD

**Issue by hylBen** [tkipf/gae#48]
Hi @tkipf ,

It seems _model.reconstructions_ includes all the positive and negative edges, no matter in training, validation or test set, whereas the labels used during training only have "1" entries on _train_edges_. It confuses me since it looks like **"the model is trained to score "1" for  _train_edges_ and "0" for _val_edges_ and _test_edges_,"** how come the validation and test accuracy are that good? Would it be better to mask out  _val_edges_ and _test_edges_ before feeding the _model.reconstructions_ to the optimizer?

Looking forward to your reply.
Thanks.

**Comment by tkipf** [comment]
This is a very good question -- this approach works as the model is
typically not fitting the training data perfectly and hence has a certain
inductive bias that allows it to generalize to the test/validation set
edges. Furthermore, the number of negative edges is usually much larger
than the number of positive edges, which further reduces the effect of
initially providing the "wrong" negative edge labels.

Masking out val- and test-edges is problematic, as this would mean one has
to know in advance which edges are in the validation and test set, which of
course does not work like this on real-world problems.

**Comment by tkipf** [comment]
Yes, it will probably work better, but most likely not by much (depending
on how large your dataset is). At the same time, this will "leak"
validation and test set information into the training set, unless you mask
out both positive and "negative" edges.

**Comment by cfpark00** [comment]
Hello, I have a perhaps related question. It seems that the model has the full graph in its GCN layer, but is trained so that the generator will match the remaining(without test,val edges) graph.
This confuses me a little bit:
1. Is it "fair" to give the model the full adjacency?
2. Is it fine train to get the remaining graph? (This is answered by this thread)

I guess both of these issues have minor impact on what actually happens?
